app/uk/gov/hmrc/serviceconfigs/persistence/DeploymentConfigSnapshotRepository.scala

```diff
@@ -66,16 +79,118 @@ class DeploymentConfigSnapshotRepository @Inject()(
+          equal("latest", true),
+          equal("deleted", false),
+        ),
+        update = set("latest", false)
+      ).toFuture().map(_ => ())
+
+  private [persistence] def removeLatestFlagForServiceInEnvironment(serviceName: String, environment: Environment): Future[Unit] =
+    collection
+      .updateMany(
+        filter = and(
+          equal("deploymentConfig.name", serviceName),
+          equal("deploymentConfig.environment", environment.asString),
+        ),
+        update = set("latest", false)
+      ).toFuture().map(_ => ())
+
+
   def populate(date: Instant): Future[Unit] = {
 
     def forEnvironment(environment: Environment): Future[Unit] = {
       for {
         deploymentConfigs <- deploymentConfigRepository.findAll(environment)
-        snapshots         =  deploymentConfigs.map(DeploymentConfigSnapshot(date, _))
-        _                 <- collection.insertMany(snapshots).toFuture()
+        latestSnapshots   <- latestSnapshotsInEnvironment(environment)
+        planOfWork        =  PlanOfWork.fromLatestSnapshotsAndCurrentDeploymentConfigs(latestSnapshots.toList, deploymentConfigs.toList, date)
+        _                 <- executePlanOfWork(planOfWork, environment)
       } yield ()
     }
 
     Environment.values.foldLeftM(())((_, env) => forEnvironment(env))
   }
+
+  def executePlanOfWork(planOfWork: PlanOfWork, environment: Environment): Future[Unit] = {
+    logger.debug(s"Processing `DeploymentConfigSnapshot`s for ${environment.asString}")
+
+    def reintroduceServiceSnapshot(deploymentConfigSnapshot: DeploymentConfigSnapshot) = {
+      logger.debug(s"Creating a snapshot for reintroduced service: $deploymentConfigSnapshot")
+      for {
+        _ <- removeLatestFlagForServiceInEnvironment(
+          deploymentConfigSnapshot.deploymentConfig.name,
+          deploymentConfigSnapshot.deploymentConfig.environment
+        )
+        _ <- collection.insertOne(deploymentConfigSnapshot).toFuture()
+      } yield ()
+    }
+
+    for {
+      _ <- if (planOfWork.snapshots.nonEmpty) removeLatestFlagForNonDeletedSnapshotsInEnvironment(environment) else Future.unit
+      _ <- collection.insertMany(planOfWork.snapshots).toFuture()
+      _ <- planOfWork.snapshotServiceReintroductions.foldLeftM(())((_, ssr) => reintroduceServiceSnapshot(ssr))
+    } yield ()
+  }
+}
+
+object DeploymentConfigSnapshotRepository {
+
+  final case class PlanOfWork(
+    snapshots: List[DeploymentConfigSnapshot],
+    snapshotServiceReintroductions: List[DeploymentConfigSnapshot]
+  )
+
+  object PlanOfWork {
+
+    def fromLatestSnapshotsAndCurrentDeploymentConfigs(
+      latestSnapshots: List[DeploymentConfigSnapshot],
+      currentDeploymentConfigs: List[DeploymentConfig],
+      date: Instant
+    ): PlanOfWork = {
+      val latestSnapshotsByNameAndEnv =
+        latestSnapshots
+          .map(s => (s.deploymentConfig.name, s.deploymentConfig.environment) -> s)
```

Comment:
or `latestSnapshots.groupBy`?

Reply:
Gives me a `Map[(String, Environment), List[DeploymentConfig]]`, i.e., doesn't take care of potential duplicates.